File: frontend/src/stores/distribution.ts
/**
 * v1.1 "Onboarding a product" §A — state for **Houston pushing stock out to the stores**.
 *
 * Same conventions as `stores/purchasing.ts`, deliberately:
 *
 *  - **Actions never throw.** On failure they set `error` and return `null`, so a sheet can put
 *    the message in a banner without wrapping every call in try/catch. `send`'s refusal is
 *    multi-line with `•` bullets naming the shortfall per item — it is kept **verbatim**, newlines
 *    and all, because reformatting it into one line is what loses the shortfalls.
 *  - `loading` covers the plan read, `busy` holds a marker while a write runs.
 *  - Writes are gated behind the warehouse store's `allowed` getter; the server enforces the same
 *    gate regardless (warehouse admin / head office only — pushing is Houston's act).
 *
 * The plan is held **per item code** rather than as one list: the sheet is opened for one item at
 * a time, from Stock, from Buying and from the Inbound receipt confirmation, and re-opening it for
 * an item already planned should not blank the screen while a request is in flight.
 */
import { defineStore } from 'pinia'
import {
  distributionApi,
  DEFAULT_COVER_DAYS,
  type DistributionLine,
  type DistributionStore,
  type PlanItem,
  type SendResult,
  type SplitMode,
  type SplitResult
} from '@/api/distribution'
import type { Priority } from '@/api/warehouse'
import { useWarehouseStore } from '@/stores/warehouse'

interface DistributionState {
  /** the enabled shops a push may address, in store-code order */
  stores: DistributionStore[]
  /** HOU-WH — where a push leaves from */
  warehouse: string | null
  /** item_code → its plan row (Houston's position plus one row per store) */
  plans: Record<string, PlanItem>
  velocityDays: number
  asOf: string | null
  /** what the last `send` created, for the confirmation */
  lastSend: SendResult | null
  loading: boolean
  busy: string | null
  error: string | null
  notice: string | null
}

function message(e: unknown): string {
  return (e as Error)?.message || 'Something went wrong'
}

export const useDistributionStore = defineStore('distribution', {
  state: (): DistributionState => ({
    stores: [],
    warehouse: null,
    plans: {},
    velocityDays: 28,
    asOf: null,
    lastSend: null,
    loading: false,
    busy: null,
    error: null,
    notice: null
  }),
  getters: {
    /** May this user push stock out of Houston? Delegates to the warehouse store. */
    allowed(): boolean {
      return useWarehouseStore().allowed
    },
    storeCount: (s): number => s.stores.length,
    planFor:
      (s) =>
      (itemCode: string): PlanItem | null =>
        s.plans[itemCode] ?? null
  },
  actions: {
    /** Load `me` once, then ask the warehouse store's `allowed` getter. Sets `error` when refused. */
    async ensureAllowed(): Promise<boolean> {
      const warehouse = useWarehouseStore()
      if (!warehouse.me && !warehouse.meError) await warehouse.loadMe()
      if (!warehouse.allowed) {
        this.error = 'Sending stock to stores is Houston’s — warehouse admin or head office only'
        return false
      }
      return true
    },
    clearNotice() {
      this.notice = null
    },
    clearError() {
      this.error = null
    },

    /** The shops a push may address. Cached — the list does not change while a sheet is open. */
    async loadStores(force = false) {
      if (this.stores.length && !force) return { stores: this.stores, count: this.stores.length, warehouse: this.warehouse || '' }
      this.error = null
      try {
        const out = await distributionApi.stores()
        this.stores = out.stores
        this.warehouse = out.warehouse
        return out
      } catch (e) {
        this.error = message(e)
        return null
      }
    },

    /** Houston's position and every store's, for one item or several. */
    async loadPlan(itemCodes: string[], boutiques?: string[] | null) {
      const codes = [...new Set((itemCodes || []).map((c) => (c || '').trim()).filter(Boolean))]
      if (!codes.length) {
        this.error = 'Choose at least one item to distribute'
        return null
      }
      this.loading = true
      this.error = null
      try {
        const out = await distributionApi.plan(codes, boutiques ?? null)
        for (const item of out.items) this.plans[item.item_code] = item
        this.stores = out.stores
        this.warehouse = out.warehouse
        this.velocityDays = out.velocity_days
        this.asOf = out.as_of
        return out
      } catch (e) {
        this.error = message(e)
        return null
      } finally {
        this.loading = false
      }
    },

    /**
     * Ask the server for an allocation. It is a **calculator, not a gate** — it will allocate more
     * than Houston has, so the sheet's footer can turn red before the send. `send` is what refuses.
     */
    async suggest(itemCode: string, qty: number, mode: SplitMode = 'even', coverDays: number | null = null, boutiques?: string[] | null): Promise<SplitResult | null> {
      this.busy = `split:${mode}`
      this.error = null
      try {
        return await distributionApi.suggest_split(itemCode, qty, mode, boutiques ?? null, coverDays ?? (mode === 'topup' ? DEFAULT_COVER_DAYS : null))
      } catch (e) {
        this.error = message(e)
        return null
      } finally {
        this.busy = null
      }
    },

    /**
     * Send it: one request + one shipment per store, created **and** approved in one action
     * (client decision 2 — the warehouse admin is both requester and approver, so a Pending step
     * would be theatre). Refused as a whole or written as a whole; never half-sent.
     */
    async send(lines: DistributionLine[], reason?: string | null, priority: Priority | string = 'Normal'): Promise<SendResult | null> {
      if (!(await this.ensureAllowed())) return null
      if (!lines?.length) {
        this.error = 'Nothing to send — choose at least one store and quantity'
        return null
      }
      this.busy = 'send'
      this.error = null
      try {
        const out = await distributionApi.send(lines, reason ?? null, priority)
        this.lastSend = out
        // the units are spoken for now: they sit in `committed` until the shipment leaves, so the
        // next plan read for this item shows less available. Refresh what we hold rather than
        // leaving a stale "available" on screen behind the confirmation.
        const codes = [...new Set(lines.map((l) => l.item_code))]
        void this.loadPlan(codes)
        this.notice = sendNotice(out)
        return out
      } catch (e) {
        // verbatim: the refusal is multi-line, one `•` per item, and each bullet names a shortfall
        this.error = message(e)
        return null
      } finally {
        this.busy = null
      }
    }
  }
})

/** "3 shipments created · 84 units to 3 stores" — the one line after a successful push. */
export function sendNotice(out: SendResult): string {
  const stores = `${out.stores} store${out.stores === 1 ? '' : 's'}`
  const units = `${out.units} unit${out.units === 1 ? '' : 's'}`
  const names = out.shipments.map((s) => s.name).join(', ')
  return `${units} on their way to ${stores} — ${names}`
}

File: maison_pos/api/distribution.py
"""Distribution API (v1.1 §A) — ``maison_pos.api.distribution.*``.

Houston pushes stock **out** to the stores. Everything here is a thin, permission-checked wrapper
over :mod:`maison_pos.distribution`; the documents themselves are still created by the existing
``maison_pos.api.shipping.create_request`` / ``approve`` pair, so a pushed shipment is an ordinary
shipment on the wall, in the pick list and on the store's Receive screen.

======================  ==========================================================================
``plan``                what Houston holds, what is committed, and every store's position
``suggest_split``       ``even`` / ``velocity`` / ``topup`` allocation helpers
``send``                create + approve one shipment per store, all or nothing
``stores``              the enabled shops a push may address (the sheet's row list)
======================  ==========================================================================

Permissions: **AWANZ Warehouse Admin** and **AWANZ Head Office** only (client decision 1 — buying
and distribution are both centralised in Houston). A store manager calling :func:`send` for their
own store is refused: pushing is Houston's act, not the store's.
"""

from __future__ import annotations

from typing import Any, Optional

import frappe

from maison_pos import distribution as dist_lib

__all__ = ["plan", "suggest_split", "send", "stores"]


@frappe.whitelist()
def stores() -> dict[str, Any]:
	"""The enabled shops a distribution may address, in store-code order.

	The head-office warehouse row is never one of them — Houston cannot push to itself.
	"""
	dist_lib.assert_distribution_admin()
	rows = dist_lib.store_rows()
	return {"stores": rows, "count": len(rows), "warehouse": dist_lib.main_warehouse()}


@frappe.whitelist()
def plan(item_codes: Any = None, boutiques: Any = None, days: Any = dist_lib.VELOCITY_DAYS) -> dict[str, Any]:
	"""Per item: HOU-WH on hand / committed / available, plus a row per store.

	``item_codes`` accepts a list, a JSON array or a comma-separated string. Each store row
	carries its on-hand, its 28-day velocity, its days of cover and whether it has ever sold the
	item — which is what makes an allocation a decision rather than a guess.
	"""
	dist_lib.assert_distribution_admin()
	return dist_lib.plan(item_codes, boutiques=boutiques, days=days)


@frappe.whitelist()
def suggest_split(item_code: str, qty: Any, mode: str = "even", boutiques: Any = None, cover_days: Any = None) -> dict[str, Any]:
	"""Allocation helper: ``even``, ``velocity`` or ``topup`` (see :mod:`maison_pos.distribution`).

	Returns a row for **every** candidate store, allocated or not, so the sheet fills all of its
	quantity boxes from one call, plus ``left_at_warehouse`` for the running footer.
	"""
	dist_lib.assert_distribution_admin()
	return dist_lib.suggest_split(item_code, qty, mode=mode, boutiques=boutiques, cover_days=cover_days)


# POST only: this one creates shipments, and a state-changing endpoint reachable by GET is a
# link away from being triggered by someone else's page (Frappe only checks CSRF on non-GET).
@frappe.whitelist(methods=["POST"])
def send(lines: Any, reason: Optional[str] = None, priority: str = "Normal") -> dict[str, Any]:
	"""Send it: one ``AWANZ Replenishment Request`` + one ``AWANZ Shipment`` per store.

	``lines = [{boutique, item_code, qty}]``. Validated in full before anything is written — a
	store that is disabled, an item that is not stock, a quantity at or below zero, or a total
	beyond what Houston actually has all refuse the **whole** distribution with the shortfall
	named per item. Nothing is left half-sent.
	"""
	dist_lib.assert_distribution_admin()
	return dist_lib.send(lines, reason=reason, priority=priority)
